### PR TITLE
🏦 Feat/logic bank balances

### PR DESCRIPTION
**Pull request by ccamel**
## 📝 Purpose

Implements #262.

## ⚙️ Implementation notes

- The arguments mode (see: [predicate description doc](https://www.swi-prolog.org/pldoc/man?section=preddesc)) are such that at call time, the arguments must be bound to a partial term (a term which may or may not be ground) satisfying some (informal) conditions (like the type and the structure). That's why the implementation is a little bit more complex than the others.
- a lot of code has been done to put the predicate under test. This work can be reused to facilitate all future developments (I hope)

## 👨‍💻 Play with it

Start the chain, as usual:

```shell
make chain-start
```

Send the following queries:

```shell
$ # Ask for all balances (only return the 1st result as configured in the module)
$ okp4d query logic ask "bank_balances(X, Y)." --output json | jq
{
  "height": "27",
  "gas_used": "9614",
  "answer": {
    "success": true,
    "has_more": true,
    "variables": [
      "X",
      "Y"
    ],
    "results": [
      {
        "substitutions": [
          {
            "variable": "Y",
            "term": {
              "name": "[uknow-999000000]",
              "arguments": []
            }
          },
          {
            "variable": "X",
            "term": {
              "name": "okp41p8u47en82gmzfm259y6z93r9qe63l25dfwwng6",
              "arguments": []
            }
          }
        ]
      }
    ]
  }
}
```

```shell
$ # Ask for all balances concerning the uknow denomination for which the balance is higher than a provided amount.
$ okp4d query logic ask "bank_balances_has_coin(A, 'uknow', V, 100000000)." --program "bank_balances_has_coin(A, D, V, S) :- bank_balances(A, R), member(D-V, R), compare(>, V, S)." --output json | jq
{
  "height": "79",
  "gas_used": "9618",
  "answer": {
    "success": true,
    "has_more": false,
    "variables": [
      "A",
      "V"
    ],
    "results": [
      {
        "substitutions": [
          {
            "variable": "A",
            "term": {
              "name": "okp41p8u47en82gmzfm259y6z93r9qe63l25dfwwng6",
              "arguments": []
            }
          },
          {
            "variable": "V",
            "term": {
              "name": "999000000",
              "arguments": []
            }
          }
        ]
      }
    ]
  }
}
```

## Acceptance Criteria

- [x] The predicate is implemented according to the described behavior and examples
- [x] The predicate passes all test cases
- [x] Any related documentation is updated

### CODE REVIEW

**Review by bdeneux — APPROVED** [review]
Very nice ! 😍 Looks good to me ☺️

File: x/logic/predicate/bank.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		}
+		bankKeeper := sdkContext.Value(types.BankKeeperContextKey).(types.BankKeeper)
+
+		bech32Addr := sdk.AccAddress(nil)
+		switch acc := env.Resolve(account).(type) {
+		case engine.Variable:
+		case engine.Atom:
+			bech32Addr, err = sdk.AccAddressFromBech32(acc.String())
+			if err != nil {
+				return engine.Error(fmt.Errorf("bank_spendable_coins/2: %w", err))
```
Comment (amimart) [inline x/logic/predicate/bank.go:45]: The error naming seems to not fit regarding the predicate name. There are other similar issues i didn't point out.

```suggestion
				return engine.Error(fmt.Errorf("bank_balances/2: %w", err))
```

**Review by amimart — CHANGES_REQUESTED** [review]
Nice one, the test structure will surely help implementing new predicates!

I just noted some inconsistencies regarding predicate names in error messages.

File: x/logic/predicate/bank.go (hunk 1 above)
Comment (ccamel) [reply x/logic/predicate/bank.go:45]: Good catch! Missed that one. :+1:

**Review by amimart — APPROVED** [review]
🚀

**Pull request merged**